src/rust-crypto/rust-crypto.ts

```diff
@@ -583,7 +600,45 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
             await this.secretStorage.store("m.cross_signing.master", crossSigningPrivateKeys.masterKey);
             await this.secretStorage.store("m.cross_signing.user_signing", crossSigningPrivateKeys.userSigningKey);
             await this.secretStorage.store("m.cross_signing.self_signing", crossSigningPrivateKeys.self_signing_key);
+
+            if (setupNewKeyBackup) {
+                await this.resetKeyBackup();
+            }
+        }
+    }
+
+    public async resetKeyBackup(): Promise<void> {
+        // check if there is already an existing backup
+        const backupInfo = await this.backupManager.setUpKeyBackup(this.signObject.bind(this));
+
+        // we want to store the private key in 4S
+        // need to check if 4S setup?
+        if (await this.secretStorageHasAESKey()) {
+            await this.secretStorage.store("m.megolm_backup.v1", backupInfo.decryptionKey.toBase64());
         }
+
+        // we can check and start async
+        this.checkKeyBackupAndEnable();
+    }
+
+    public async signObject<T extends ISignableObject & object>(obj: T): Promise<void> {
+        const sigs = new Map(Object.entries(obj.signatures || {}));
+        const unsigned = obj.unsigned;
+
+        delete obj.signatures;
+        delete obj.unsigned;
+
+        const userSignatures = sigs.get(this.userId) || {};
+
+        const canonalizedJson = anotherjson.stringify(obj);
+        const signatures: RustSdkCryptoJs.Signatures = await this.olmMachine.sign(canonalizedJson);
+
+        const map = JSON.parse(signatures.asJSON());
+
+        sigs.set(this.userId, { ...userSignatures, ...map[this.userId] });
```

Comment:
rather than going via JSON:

```suggestion
        const userSignatures = signatures.get(new RustSdkCryptoJs.UserId(this.userId));
        sigs.set(this.userId, { 
            ...(sigs.get(this.userId) ?? {}),
            ...Object.fromEntries(userSignatures.entries()),
        });
```

Reply:
The wasm api is quite unpractical. Looks simpler in Json to me

Reply:
ok